Clarified Sync Error Message

puranjayK opened the pull request:
## Pull Request template

## Purpose / Description
To clarify the sync error message to the user

## Fixes
Fixes #10000 

## Approach
Used the progress value of upload and download to clarify the user about when the sync stopped.

## Screenshot
<img src="https://user-images.githubusercontent.com/85157266/147289008-54394d27-f597-4c1f-909e-d160fe51bb19.png" width = 250/>


## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

Arthur-Milchior reviewed (commented):
Hi,

Thanks a lot for your first contribution. That's great to see indication that some down/up load took place.

I fear this is quite far from what I imagined. Indeed, for a user who is not used to computer internals, they still would not know what to do, nor what did occur. Let's concentrate on both part separately. 

As I suggested in the issue, we can explain to the user they can retry; maybe they need a stable internet connection, or to wait because ankiweb is buggy, or something. The action, even if it fail, was not useless, and even if it took many tries and failed in the way, it's still worth syncing again and again. But the user do not know that with the information they have, so we need to let them know it makes sens.
To show it actually makes sens, I believe the best way is to tell them how many medias were downloaded or uploaded. So they know something did really occur. Ideally they should also know what remains to do, so that they can judge the progress. After all, if they can only upload two medias at a time and they have 2000 medias to upload, something is very wrong. If they have 2000 medias to upload and they can do 500 before their connection broke, it will "only" take four tries, which is more reasonnable for them.
Even better, I think, would be to only mention download/upload depending on when the process failed. So they get only the relevant informations. After all if they always see (0 download) but it is only because there is actually nothing to download, this information is useless and scary. If they always see (0 upload) because upload only occur when downloads are done, this information is also useless (we didn't even try to upload anything) and scary.


Most of the informations are either available in the media syncer. And you can add more properties if necessary to store all of the data you need to have access to. You'll need to look at MediaSyncer's sync function to look at when upload/download did occur to take care of it. And also log the number of expected download and upload.

Sorry if it's more complex than you expected. I still believe it can be a good first contribution, but I also fear you'll need to do a little more work on it first.

puranjayK commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -578,6 +578,8 @@ public void publishProgress(String message) {
     @SuppressWarnings("deprecation") // #7108: AsyncTask
     public void publishProgress(int id, long up, long down) {
         super.publishProgress(id, up, down);
+        kBytesSent =up/1024;
```
Done.

puranjayK commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -578,6 +578,8 @@ public void publishProgress(String message) {
     @SuppressWarnings("deprecation") // #7108: AsyncTask
     public void publishProgress(int id, long up, long down) {
         super.publishProgress(id, up, down);
+        kBytesSent =up/1024;
```
Done.

puranjayK commented:
> Hi,
> 
> Thanks a lot for your first contribution. That's great to see indication that some down/up load took place.
> 
> I fear this is quite far from what I imagined. Indeed, for a user who is not used to computer internals, they still would not know what to do, nor what did occur. Let's concentrate on both part separately.
> 
> As I suggested in the issue, we can explain to the user they can retry; maybe they need a stable internet connection, or to wait because ankiweb is buggy, or something. The action, even if it fail, was not useless, and even if it took many tries and failed in the way, it's still worth syncing again and again. But the user do not know that with the information they have, so we need to let them know it makes sens. To show it actually makes sens, I believe the best way is to tell them how many medias were downloaded or uploaded. So they know something did really occur. Ideally they should also know what remains to do, so that they can judge the progress. After all, if they can only upload two medias at a time and they have 2000 medias to upload, something is very wrong. If they have 2000 medias to upload and they can do 500 before their connection broke, it will "only" take four tries, which is more reasonnable for them. Even better, I think, would be to only mention download/upload depending on when the process failed. So they get only the relevant informations. After all if they always see (0 download) but it is only because there is actually nothing to download, this information is useless and scary. If they always see (0 upload) because upload only occur when downloads are done, this information is also useless (we didn't even try to upload anything) and scary.
> 
> Most of the informations are either available in the media syncer. And you can add more properties if necessary to store all of the data you need to have access to. You'll need to look at MediaSyncer's sync function to look at when upload/download did occur to take care of it. And also log the number of expected download and upload.
> 
> Sorry if it's more complex than you expected. I still believe it can be a good first contribution, but I also fear you'll need to do a little more work on it first.

Thank you so much for the clarification. I'll definitely look into it. 
I was able to reproduce this issue only when the upload / download was almost / totally completed. It would be helpful if you could tell on how should I reproduce it in other situations as well (if any).

Arthur-Milchior commented:
I believe that the easiest way to really stop syncing is:
* add a lot of heavy media, so that trying will take dozens of second
* during the media sync, move to plane mode to remove internet access

An even easier way to do throw an exception during media sync is to hard code it. Modify the MediaSync's sync method, and throw an exception after one or two media downloaded/uploaded. (of course, this should not be comitted)

But anyway, I don't believe that it's really important to test the code as it is right now. First show the information that are useful for the user in case of problem, as I described above*. Once the feature is implemented, it'll make more sent to actually look at whether we get the right error message.

*Or suggest something even better, or disagree and explain why my idea is not as good as I think it is. I am never asking you to just obey, all contributors are welcome to have opinions and disagree. We just want to know why you do so. In this case, if you think your current message is better than my suggestion, I'd love to know why.

ShridharGoel commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -482,7 +482,7 @@ private Payload doInBackgroundSync(Payload data) {
                     ret = mediaClient.sync();
                     if (ret == null || ret.first == null) {
-                        mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error);
+                        mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error)+ "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_up_down_size,kBytesSent,kBytesReceived);
```
```suggestion
                        mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_up_down_size, kBytesSent, kBytesReceived);
```

puranjayK replied:
These variables were not needed, hence I deleted them

ShridharGoel commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -507,7 +507,7 @@ private Payload doInBackgroundSync(Payload data) {
                         data.result = new Object[]{e};
                     }
-                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage();
+                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage() + "\n\n"+ AnkiDroidApp.getAppResources().getString(R.string.sync_up_down_size,kBytesSent,kBytesReceived);
```
```suggestion
                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage() + "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_up_down_size, kBytesSent, kBytesReceived);
```

puranjayK replied:
Done.

ShridharGoel reviewed (changes requested):
Small formatting changes

puranjayK commented:
> Hi,
> 
> Thanks a lot for your first contribution. That's great to see indication that some down/up load took place.
> 
> I fear this is quite far from what I imagined. Indeed, for a user who is not used to computer internals, they still would not know what to do, nor what did occur. Let's concentrate on both part separately.
> 
> As I suggested in the issue, we can explain to the user they can retry; maybe they need a stable internet connection, or to wait because ankiweb is buggy, or something. The action, even if it fail, was not useless, and even if it took many tries and failed in the way, it's still worth syncing again and again. But the user do not know that with the information they have, so we need to let them know it makes sens. To show it actually makes sens, I believe the best way is to tell them how many medias were downloaded or uploaded. So they know something did really occur. Ideally they should also know what remains to do, so that they can judge the progress. After all, if they can only upload two medias at a time and they have 2000 medias to upload, something is very wrong. If they have 2000 medias to upload and they can do 500 before their connection broke, it will "only" take four tries, which is more reasonnable for them. Even better, I think, would be to only mention download/upload depending on when the process failed. So they get only the relevant informations. After all if they always see (0 download) but it is only because there is actually nothing to download, this information is useless and scary. If they always see (0 upload) because upload only occur when downloads are done, this information is also useless (we didn't even try to upload anything) and scary.
> 
> Most of the informations are either available in the media syncer. And you can add more properties if necessary to store all of the data you need to have access to. You'll need to look at MediaSyncer's sync function to look at when upload/download did occur to take care of it. And also log the number of expected download and upload.
> 
> Sorry if it's more complex than you expected. I still believe it can be a good first contribution, but I also fear you'll need to do a little more work on it first.

I had 45 media files on AnkiWeb server. I reinstalled Ankidroid (delete the folder in root) and then logged in and synced.
1                                    |                2                    |          3
:-------------------------:|:-------------------------:|:-------------------------:
<img src ="https://user-images.githubusercontent.com/85157266/149774130-6e6997e3-bc58-4db6-9caa-53705d2c25cf.jpeg" width =300/> | <img src= "https://user-images.githubusercontent.com/85157266/149774120-7216ebae-0d3d-4794-b378-39c89cb29e0f.jpeg" width = 300/> | <img src ="https://user-images.githubusercontent.com/85157266/149774134-5617848b-cf35-4421-9845-657af9158eb6.jpeg" width = 300/>

1.  I switched off all internet access in the mid of the sync (around 25 media files were downloaded).
2.  Then I enabled internet connection and synced again. while adding 7 media files. On syncing the remaining 20 media files were downloaded and then the upload began. In the mid of the upload, I turned off internet again (around 5 files had been uploaded by that point).
3.  Finally I switched on the internet and synced again and the remaining 2 files got uploaded.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -77,6 +77,16 @@
      */
     private final PowerManager.WakeLock mWakeLock;
+    private long kBytesReceived, kBytesSent;
+    private long filesDownloaded=0;
```
https://github.com/ankidroid/Anki-Android/wiki/Code-style

```suggestion
    private long filesDownloaded = 0;
```

puranjayK replied:
Done

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -77,6 +77,16 @@
      */
     private final PowerManager.WakeLock mWakeLock;
+    private long kBytesReceived, kBytesSent;
+    private long filesDownloaded=0;
```
Can these be moved to `MediaSyncer`, they're not a responsibility of the connection.

puranjayK replied:
I'll need their value in Connection.java because the value is displayed in a connection error

david-allison-1 replied:
I'm fairly sure these can be moved, you should have access to `MediaSyncer` from the catch block

puranjayK replied:
Yup, got it

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -77,6 +77,16 @@
      * the process will be paused and the sync can fail due to timing conflicts with AnkiWeb.
      */
     private final PowerManager.WakeLock mWakeLock;
+    private long kBytesReceived, kBytesSent;
```
Should be split into two variables

puranjayK replied:
I have removed these as they aren't needed anymore

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -507,7 +517,7 @@ private Payload doInBackgroundSync(Payload data) {
                         data.result = new Object[]{e};
                     }
-                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage();
+                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage() + "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_media_downloaded_count,filesDownloaded)  + "\n\n" + "Uploaded "+filesUploaded+ " media files";
```
```suggestion
                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage() + "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_media_downloaded_count, filesDownloaded)  + "\n\n" + "Uploaded "+ filesUploaded + " media files";
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -507,7 +517,7 @@ private Payload doInBackgroundSync(Payload data) {
                         data.result = new Object[]{e};
                     }
-                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage();
+                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage() + "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_media_downloaded_count,filesDownloaded)  + "\n\n" + "Uploaded "+filesUploaded+ " media files";
```
This needs to be extracted to a string

puranjayK replied:
Done.

david-allison-1 reviewed (changes requested):
Few comments, nothing major

david-allison-1 reviewed (changes requested):
I still feel the variables can be moved to `MediaSyncer`

puranjayK commented:
> I still feel the variables can be moved to `MediaSyncer`

PTAL, I have made the changes

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/libanki/sync/MediaSyncer.java:
```diff
@@ -197,6 +206,7 @@ public MediaSyncer(Collection col, RemoteMediaServer server, Connection con) {
                                                        AnkiDroidApp.getAppResources().getString(R.string.sync_media_changes_count), toSend));
 
                     JSONArray changes = mServer.uploadChanges(zip);
+                    mUploadCount = changes.getInt(0);
```
I'd have expected `mUploadCount += processedCnt` here. Is there a reason we're resetting the variable?

puranjayK replied:
Yup this would make more sense.

david-allison-1 reviewed (commented):
One question.

Could you update the screenshot on the PR? It looks out of date

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/Connection.java:
```diff
@@ -507,7 +507,9 @@ private Payload doInBackgroundSync(Payload data) {
-                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage();
+                    mediaError = AnkiDroidApp.getAppResources().getString(R.string.sync_media_error) + "\n\n" + e.getLocalizedMessage()
+                            + "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_media_downloaded_count, mediaClient.getDownloadCount())
+                            + "\n\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_media_uploaded_count, mediaClient.getUploadCount());
```
I think one `\n` would be better here

```suggestion
                            + "\n" + AnkiDroidApp.getAppResources().getString(R.string.sync_media_uploaded_count, mediaClient.getUploadCount());
```

puranjayK replied:
@david-allison-1 Can you give me a feedback on this

david-allison-1 replied:
Sounds good to me

puranjayK commented:
1                                    |                2                    |          3
:-------------------------:|:-------------------------:|:-------------------------:
 <img src ="https://user-images.githubusercontent.com/85157266/150950856-1bda066b-bb26-41f2-ab70-7af0129715b8.jpeg" width =300/>  | <img src ="https://user-images.githubusercontent.com/85157266/150950846-e65b4655-9c80-4ec1-9d02-deeea5b72c6c.jpeg" width =300/> | <img src ="https://user-images.githubusercontent.com/85157266/150950862-a945dfe2-bab6-4795-9f08-1f0807b15db5.jpeg" width =300/>

1. In the mid of the upload, I turned off internet again (around 4 files had been uploaded by that point). Finally I switched on the internet and synced again and the remaining  files got uploaded.
2. I switched off all internet access in the mid of the sync (around 39 media files were downloaded).
3. Then I enabled internet connection and synced again. On syncing the remaining media files were downloaded.

david-allison-1 commented:
Hi, the screenshots don't seem to match the review comments I made.

puranjayK commented:
> Hi, the screenshots don't seem to match the review comments I made.

Sorry, I missed the review comments. I'll make the updates and post the screenshots

puranjayK commented:
1                                    |                2                    |          3
:-------------------------:|:-------------------------:|:-------------------------:
<img src ="https://user-images.githubusercontent.com/85157266/152633588-a80f61a7-e54a-4379-b136-6ea20c3c8ee9.jpeg" width =300/> | <img src= "https://user-images.githubusercontent.com/85157266/152633584-0aff4996-a3b3-4145-94d2-97d0541c7384.jpeg" width = 300/> | <img src ="https://user-images.githubusercontent.com/85157266/152633577-ee648461-e062-4b08-94ca-b8326555d361.jpeg" width = 300/>

1.  I switched off all internet access at the start of the sync (no media files were downloaded / uploaded).
2.  Then I enabled internet connection and synced again. On syncing 11 media files were downloaded and then I turned off internet again.
3.  Finally I switched on the internet , while adding 7 media files and synced again and the remaining files got downloaded and the upload started. I switched off the internet when 3 media files were uploaded.

david-allison-1 reviewed (changes requested):
We no longer show the error message on the dialog

puranjayK commented:
> This no longer shows the error message on the dialog

I thought you asked me to remove that completely

puranjayK commented:
1                                    |                2                    |       
:-------------------------:|:-------------------------:
<img src ="https://user-images.githubusercontent.com/85157266/152634603-566f2871-f662-46bb-bcd2-1bb3c6f10be9.jpeg" width =300/> | <img src= "https://user-images.githubusercontent.com/85157266/152634612-43c48853-03f7-4c7f-899f-2073775e03f0.jpeg" width = 300/> 

@david-allison-1 PTAL

david-allison-1 commented:
> 1                                    |                2                    |       
> :-------------------------:|:-------------------------:
> <img src ="https://user-images.githubusercontent.com/85157266/152634603-566f2871-f662-46bb-bcd2-1bb3c6f10be9.jpeg" width =300/> | <img src= "https://user-images.githubusercontent.com/85157266/152634612-43c48853-03f7-4c7f-899f-2073775e03f0.jpeg" width = 300/> 
> 
> @david-allison-1 PTAL

Could you switch the error message to be last in 2.

Looks good aside from that

puranjayK commented:
> Could you switch the error message to be last in **2** and add a single newline between the message, upload and download strings (not a gap, just so each string starts in a new line)
> 
> Looks good aside from that
<img src="https://user-images.githubusercontent.com/85157266/152644913-ab74fd70-53a4-4e37-a8e7-eecbe1461e77.jpeg" width =300/>

@david-allison-1 PTAL

Arthur-Milchior commented:
@puranjayK There is a conflict that need to be resolved in MediaSyncer.java, sorry. Could you please resolve it and push again.

The test failed, but I read
> Platform error occurred when running the UTP test suite

It seems the issue was with the test system and not your code itself, so hopefully, once the test run again with the conflict merged, it'll work.

Thanks a lot for all of those updates. I believe all of those message will be really clearer!

puranjayK commented:
> @puranjayK There is a conflict that need to be resolved in MediaSyncer.java, sorry. Could you please resolve it and push again.
> 
> The test failed, but I read
> 
> > Platform error occurred when running the UTP test suite
> 
> It seems the issue was with the test system and not your code itself, so hopefully, once the test run again with the conflict merged, it'll work.
> 
> Thanks a lot for all of those updates. I believe all of those message will be really clearer!

<img src ="https://user-images.githubusercontent.com/85157266/153703291-35523d9f-cb98-4af9-ba39-8c57126770ed.jpeg" width = 250/>  <img src = "https://user-images.githubusercontent.com/85157266/153703293-f5b76901-9eac-491e-af98-71990559e327.jpeg" width = 250/>

@Arthur-Milchior PTAL. 
- [x] Fixed the merge conflicts
- [x] Rebuilt the project
- [x] Tested the error and uploaded the screenshot for the same

david-allison-1 commented on AnkiDroid/src/main/res/values/04-network.xml:
```diff
@@ -81,6 +81,7 @@
     <string name="sync_media_find">Finding changed media…</string>
     <string name="sync_media_no_changes">No changes to media files</string>
     <string name="sync_media_success">Media synced</string>
+    <string name="sync_media_partial">Media sync partially completed: %d media files downloaded.\n%d media files uploaded.</string>
```
```suggestion
    <string name="sync_media_partial">Media sync partially completed:\n%d media files downloaded.\n%d media files uploaded.</string>
```

david-allison-1 reviewed (changes requested):
Sounds like we're there after this one!

david-allison-1 reviewed (approved):
LGTM

Arthur-Milchior commented:
I assume we squash everything. Or do you want to clean the PR up and do a sequence of atomic commits? Given it's only 16 lines, i don't believe it's useful

david-allison commented:
squash + manually cleanup commit message seems like the best

Arthur-Milchior commented:
> This branch cannot be rebased due to conflicts

Sorry, it seems it needs you manually remerge it again

Arthur-Milchior commented:
Oh actually, you'll need to squash it all yourself, because the reason it's blocking, I believe, is because github can't deal with intermediate commits here
